### DISCUSSION TITLE
Proposal : IEnumerable iteration performance enhancement

### DISCUSSION THREAD

**Discussion by NutzNK** [dotnet/csharplang#3138] [Language Ideas]
We have performance issue on IEnumerable interface.
because  IEnumerable take several times slower than array iteration.

problem is when collection has been casted to IEnumerable<T> or return IEnumerable<T> by some function, foreach iteration will iterate IEnumerator<T> interface from method "MoveNext" and property "Current", these methods and property cannot do inlining and there are overhead for 2 methods calling.

```C#
//This is foreach loop

foreach(var value in object)
{
  //code in loop here
}

//This is standard "Foreach" equivalent code:
using(var enumerator = object.GetEnumerator())
{
  while(enumerator.MoveNext())
  {
    var value = enumerator.Current;
    //code in loop here
  }
}
```

I would suggest supported equivalent code for "Foreach" like this

```C#
//This is proposal "Foreach" equivalent code:
using(var enumerator = object.GetEnumerator())
{
  while(enumerator.TryMoveNext(out var value))
  {
    //code in loop here
  }
}
```
with this optimized enumerator
```C#
//This is optimized enumerator
public interface IEnumerator : IDisposable
{
  bool TryGetNext(out object value);
}

public interface IEnumerator<out T> : IEnumerator 
{
  bool TryGetNext(out T value);
}
```

This is example of collection class with optimized enumerator

```C#
public class OptimizedList<T> : IEnumerable<T>
{
    private const int defaultCapacity = 4;
    private T[] items;
    private int size;

    public OptimizedList()
    {
        items = new T[defaultCapacity];
    }

    public OptimizedList(T[] source)
    {
        int size = source.Length;
        items = new T[size];
        Array.Copy(source, 0, items, 0, size);
        this.size = size;
    }
    public int Size => size;

    public T this[int index]
    {
        get
        {
            if((uint)index >= (uint)size)
                ThrowIndexOutOfRange();
            return items[index];
        }
        set
        {
            if ((uint)index >= (uint)size)
                ThrowIndexOutOfRange();
            items[index] = value;
        }
    }

    public void Add(T item)
    {
        int size = this.size;
        if (size >= items.Length)
        {
            int newCapacity = size * 2;
            if(newCapacity < 0)
            {
                if (size < int.MaxValue)
                    newCapacity = int.MaxValue;
                else
                    ThrowExceedLimitedSize();
            }
            Array.Resize(ref items, newCapacity);
        }
        items[size] = item;
        this.size = size + 1;
    }

    public void Insert(T item, int index)
    {
        int size = this.size;
        if ((uint)index >= (uint)size)
            ThrowIndexOutOfRange();
        if (size >= items.Length)
        {
            int newCapacity = size * 2;
            if (newCapacity < 0)
            {
                if (size < int.MaxValue)
                    newCapacity = int.MaxValue;
                else
                    ThrowExceedLimitedSize();
            }
            Array.Resize(ref items, newCapacity);
        }
        Array.Copy(items, index, items, index + 1, size - index);
        items[index] = item;
        this.size = size + 1;
    }

    public bool RemoveFirst(T item)
    {
        T[] items = this.items;
        for(int i = 0, size = this.size; i < size; ++i)
        {
            if(EqualityComparer<T>.Default.Equals(item, items[i]))
            {
                --size;
                if(i < size)
                {
                    Array.Copy(items, i + 1, items, i, size - i);
                }
                items[size] = default;
                this.size = size;
                return true;
            }
        }
        return false;
    }

    public bool RemoveLast(T item)
    {
        T[] items = this.items;
        for (int size = this.size, i = size - 1; i > 0; --i)
        {
            if (EqualityComparer<T>.Default.Equals(item, items[i]))
            {
                --size;
                if (i < size)
                {
                    Array.Copy(items, i + 1, items, i, size - i);
                }
                items[size] = default;
                this.size = size;
                return true;
            }
        }
        return false;
    }

    public bool RemoveAll(T item)
    {
        bool isRemoved = false;
        T[] items = this.items;
        for (int i = 0, size = this.size; i < size; ++i)
        {
            if (EqualityComparer<T>.Default.Equals(item, items[i]))
            {
                --size;
                if (i < size)
                {
                    Array.Copy(items, i + 1, items, i, size - i);
                }
                items[size] = default;
                this.size = size;
                isRemoved = true;
            }
        }
        return isRemoved;
    }

    public ValueEnumerator GetEnumerator()
    {
        return new ValueEnumerator(this);
    }

    private static void ThrowIndexOutOfRange()
    {
        throw new IndexOutOfRangeException("Access fault, Index is out of range.");
    }

    private static void ThrowExceedLimitedSize()
    {
        throw new OverflowException("Addition fault, Size is already reached the limited.");
    }

    IEnumerator<T> IEnumerable<T>.GetEnumerator()
    {
        return new Enumerator(this);
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return new Enumerator(this);
    }

    public ref struct ValueEnumerator
    {
        private T[] items;
        private int index, size;

        public ValueEnumerator(OptimizedList<T> source)
        {
            index = 0;
            items = source.items;
            size = source.size;
        }

        public void Dispose() { }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public bool TryGetNext(out T next)
        {
            int i = index;
            if (i < size)
            {
                index = i + 1;
                next = items[i];
                return true;
            }
            else
            {
                next = default;
                return false;
            }
        }
    }

    public sealed class Enumerator : IEnumerator<T>
    {
        private T[] items;
        private int index, size;

        public Enumerator(OptimizedList<T> source)
        {
            index = 0;
            items = source.items;
            size = source.size;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public bool TryGetNext(out T next)
        {
            int i = index;
            if (i < size)
            {
                index = i + 1;
                next = items[i];
                return true;
            }
            else
            {
                next = default;
                return false;
            }
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public bool TryGetNext(out object next)
        {
            int i = index;
            if (i < size)
            {
                index = i + 1;
                next = items[i];
                return true;
            }
            else
            {
                next = default;
                return false;
            }
        }
    }
}
```

this class iteration performance is close to array iteration and better than standard IEnumerator when it casted to IEnumerable(<T>).

but the next problem is this optimized enumerator interface contains method with "out" parameter(s) that mean this interface is not support co-variant type yet.
then I would suggest to co-variant type should support type that contains method with "out" parameter(s).

These are test performance on .NET core 3.1 with 1,000,000 loop of int accumulation.

Array_For => Min : 0.495 msec, Average : 0.530 msec, Median : 0.519 msec.
Array_Foreach => Min : 0.997 msec, Average : 1.082 msec, Median : 1.035 msec.
IEnumerable(Array)_Foreach => Min : 3.267 msec, Average : 4.115 msec, Median : 4.135 msec.
List_Foreach => Min : 2.247 msec, Average : 2.337 msec, Median : 2.300 msec.
IEnumerable(List)_Foreach => Min : 5.397 msec, Average : 5.693 msec, Median : 5.669 msec.
**OptimizedList_Foreach => Min : 0.495 msec, Average : 0.528 msec, Median : 0.509 msec.
IEnumerable(OptimizedList)_Foreach => Min : 1.979 msec, Average : 2.090 msec, Median : 2.025 msec.**

**Comment by HaloFour** [comment]
You could solve the covariance issue by changing the signature to return `T` and have an `out bool` parameter indicating success.  I believe this pattern has been discussed before and found to be confusing.  C# supports pattern-based `struct` enumerators which largely eliminates the performance issues.

**Comment by YairHalberstadt** [comment]
I would suggest using BenchmarkDotNet for your benchmarks. In my opinion home grown benchmarking Frameworks are useless. 

As it happens your benchmark results are clearly impossible:

1. Array_for and Array_foreach have identical performance.

2. Your optimised foreach still can't be inlined, and still has one interface call per iteration. There is no way it is faster than List_foreach which uses an optimised struct enumerator which can be inlined completely. The only possibility I can think of, is that List checks to make sure it's not modified whilst enumerating. This adds overhead. Since you aren't checking this isn't a fair comparison.

**Comment by NutzNK** [comment]
@HaloFour Yes, I know that, that why this proposal have request for making covariance type support out parameter

As I know (may be wrong, if wrong please clarify) Actually CPU instruction for this 2 method is same(refer from assembly language and c language)
1. bool MoveNext(){};
2. void MoveNext(out bool result){};

but I'm not sure is C# compiler make it same or not.
if it's made in the same way or not much different it should be able to do.

And Yes, I know this proposal has been discussed before but this proposal has different coding detail.

and Yes, struct enumeration can eliminate performance issue, I used it too, but struct has limitation for fast productivity and good scalability language like C#, Java, Python, ….
you cannot create only 1 struct that can use for all enumerable types.

this proposal is concentrate on IEnumerator interface that can give better performance that current one.

**Comment by NutzNK** [comment]
@YairHalberstadt normally I test it by my own benchmarking library, it will take shorter time while benchmarking and coding but this is result from BenchmarkDotNet.

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18363
Intel Core i7-8750H CPU 2.20GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.1.101
  [Host]     : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT
  DefaultJob : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT


|                            Method |       Mean |    Error |   StdDev |
|---------------------------------- |-----------:|---------:|---------:|
|                         Array_For |   518.4 us |  8.55 us |  7.58 us |
|                     Array_Foreach |   442.4 us |  1.05 us |  0.98 us |
|         Array_IEnumerable_Foreach | 3,503.7 us |  6.25 us |  5.85 us |
|                      Foreach_List | 2,029.0 us | 29.90 us | 26.51 us |
|          Foreach_List_IEnumerable | 5,514.4 us | 16.54 us | 15.47 us |
|           OptimizedList_Iteration |   759.7 us | 14.51 us | 14.25 us |
| OptimizedList_Interface_Iteration | 2,251.6 us | 11.25 us | 10.52 us |

This is code for benchmarking
```C#
namespace Benchmark
{
    public class Tester
    {
        const int size = 1_000_000;
        int[] array = new int[size];
        List<int> list;
        OptimizedList<int> oList;

        [GlobalSetup]
        public void Array_Setup()
        {
            for (int i = 0, len = size; i < len; ++i)
            {
                array[i] = i;
            }
            list = new List<int>(array);
            oList = new OptimizedList<int>(array);
        }

        [Benchmark]
        public void Array_For()
        {
            long result = 0;
            var array = this.array;
            for (var i = 0; i < size; ++i)
            {
                result += array[i];
            }
        }

        [Benchmark]
        public void Array_Foreach()
        {
            long result = 0;

            foreach (var item in array)
            {
                result += item;
            }
        }

        [Benchmark]
        public void Array_IEnumerable_Foreach()
        {
            long result = 0;
            var e = array.AsEnumerable();

            foreach (var item in e)
            {
                result += item;
            }
        }

        [Benchmark]
        public void Foreach_List()
        {
            double value = 0;

            foreach (var v in list)
            {
                value += v;
            }
        }

        [Benchmark]
        public void Foreach_List_IEnumerable()
        {
            int value = 0;
            IEnumerable<int> e = list;

            foreach (var v in e)
            {
                value += v;
            }
        }

        [Benchmark]
        public void OptimizedList_Iteration()
        {
            long result = 0;

            using (var e = oList.GetEnumerator())
            {
                while (e.TryGetNext(out int value))
                {
                    result += value;
                }
            }
        }

        [Benchmark]
        public void OptimizedList_Interface_Iteration()
        {
            long result = 0;
            IFastEnumerable<int> fe = oList;

            using (var e = fe.GetEnumerator())
            {
                while (e.TryGetNext(out int value))
                {
                    result += value;
                }
            }
        }
    }
}
```

Yes, This optimized interface still cannot be inlined but it take less virtual method calling for each iteration, so it take less overhead, this optimized IEnumeration call only 1 virtual method(TryGetNext) for each iteration but current IEnumerator call 2 virtual methods(MoveNext and get_Current) for each iteration.
That is reason why it can get better performance.

**Comment by YairHalberstadt** [comment]
@NutzNK 

Thanks for that - as you can see the results are significantly different compared to the previous results, and make much more sense.

However you're not comparing apples to apples here. The list iterator has some logic to make sure it doesn't update the list during iteration. That is why it's so much slower than array. Your enumerator doesn't do that, so isn't comparable to the list enumerator.

**Comment by YairHalberstadt** [comment]
Also, in some cases you're adding to a long, in others an int, and in one a double. They have to be consistent.

**Comment by NutzNK** [comment]
@YairHalberstadt 
1. sorry for my fault in accumulator type int, double and long, below test has been fixed.

2. Below test OptimizedList<T> enumerator has been implemented version checking, so function should be the same as List<T>.Enumerator but OptimizedList<T> still faster than List<T> approx. 2 times on enumerator iteration and 2.6 times on enumerator interface iteration and also faster than array enumerator interface iteration.

3. this is result of new test.

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18363
Intel Core i7-8750H CPU 2.20GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.1.101
  [Host]     : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT
  DefaultJob : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT


|                            Method |       Mean |    Error |   StdDev |
|---------------------------------- |-----------:|---------:|---------:|
|                         Array_For |   509.0 us | 10.00 us |  8.87 us |
|                     Array_Foreach |   462.1 us |  6.05 us |  5.05 us |
|         Array_IEnumerable_Foreach | 3,930.8 us |  5.06 us |  4.22 us |
|                      Foreach_List | 1,778.4 us |  5.01 us |  4.69 us |
|          Foreach_List_IEnumerable | 5,723.6 us | 98.27 us | 91.92 us |
|           OptimizedList_Iteration |   892.3 us |  9.34 us |  8.74 us |
| OptimizedList_Interface_Iteration | 2,226.5 us |  7.75 us |  7.25 us |

4. This is Benchmark class
```C#
namespace Benchmark
{
    public class Tester
    {
        const int size = 1_000_000;
        int[] array = new int[size];
        List<int> list;
        OptimizedList<int> oList;

        [GlobalSetup]
        public void Array_Setup()
        {
            for (int i = 0, len = size; i < len; ++i)
            {
                array[i] = i;
            }
            list = new List<int>(array);
            oList = new OptimizedList<int>(array);
        }

        [Benchmark]
        public void Array_For()
        {
            long result = 0;
            var array = this.array;
            for (var i = 0; i < size; ++i)
            {
                result += array[i];
            }
        }

        [Benchmark]
        public void Array_Foreach()
        {
            long result = 0;

            foreach (var item in array)
            {
                result += item;
            }
        }

        [Benchmark]
        public void Array_IEnumerable_Foreach()
        {
            long result = 0;
            var e = array.AsEnumerable();

            foreach (var item in e)
            {
                result += item;
            }
        }

        [Benchmark]
        public void Foreach_List()
        {
            long value = 0;

            foreach (var v in list)
            {
                value += v;
            }
        }

        [Benchmark]
        public void Foreach_List_IEnumerable()
        {
            long value = 0;
            IEnumerable<int> e = list;

            foreach (var v in e)
            {
                value += v;
            }
        }

        [Benchmark]
        public void OptimizedList_Iteration()
        {
            long result = 0;

            using (var e = oList.GetEnumerator())
            {
                while (e.TryGetNext(out int value))
                {
                    result += value;
                }
            }
        }

        [Benchmark]
        public void OptimizedList_Interface_Iteration()
        {
            long result = 0;
            IFastEnumerable<int> fe = oList;

            using (var e = fe.GetEnumerator())
            {
                while (e.TryGetNext(out int value))
                {
                    result += value;
                }
            }
        }
    }
}
```

5. This is class of OptimizedList<T>
```C#
namespace ExtremeCore.Collections
{
    public class OptimizedList<T> : IFastEnumerable<T>
    {
        private const int defaultCapacity = 4;
        private T[] items;
        private int size;
        private int version;

        public OptimizedList()
        {
            items = new T[defaultCapacity];
        }

        public OptimizedList(T[] source)
        {
            int size = source.Length;
            items = new T[size];
            Array.Copy(source, 0, items, 0, size);
            this.size = size;
        }

        public int Size => size;

        public T this[int index]
        {
            get
            {
                if ((uint)index >= (uint)size)
                    ThrowIndexOutOfRange();
                return items[index];
            }
            set
            {
                if ((uint)index >= (uint)size)
                    ThrowIndexOutOfRange();
                items[index] = value;
                ++version;
            }
        }

        public void Add(T item)
        {
            int size = this.size;
            if (size >= items.Length)
            {
                int newCapacity = size * 2;
                if (newCapacity < 0)
                {
                    if (size < int.MaxValue)
                        newCapacity = int.MaxValue;
                    else
                        ThrowExceedLimitedSize();
                }
                Array.Resize(ref items, newCapacity);
            }
            items[size] = item;
            this.size = size + 1;
            ++version;
        }

        public void InsertAt(T item, int index)
        {
            int size = this.size;
            if ((uint)index >= (uint)size)
                ThrowIndexOutOfRange();
            if (size >= items.Length)
            {
                int newCapacity = size * 2;
                if (newCapacity < 0)
                {
                    if (size < int.MaxValue)
                        newCapacity = int.MaxValue;
                    else
                        ThrowExceedLimitedSize();
                }
                Array.Resize(ref items, newCapacity);
            }
            Array.Copy(items, index, items, index + 1, size - index);
            items[index] = item;
            this.size = size + 1;
            ++version;
        }

        public bool RemoveFirst(T item)
        {
            T[] items = this.items;
            for (int i = 0, size = this.size; i < size; ++i)
            {
                if (EqualityComparer<T>.Default.Equals(item, items[i]))
                {
                    --size;
                    if (i < size)
                    {
                        Array.Copy(items, i + 1, items, i, size - i);
                    }
                    items[size] = default;
                    this.size = size;
                    ++version;
                    return true;
                }
            }
            return false;
        }

        public bool RemoveLast(T item)
        {
            T[] items = this.items;
            for (int size = this.size, i = size - 1; i > 0; --i)
            {
                if (EqualityComparer<T>.Default.Equals(item, items[i]))
                {
                    --size;
                    if (i < size)
                    {
                        Array.Copy(items, i + 1, items, i, size - i);
                    }
                    items[size] = default;
                    this.size = size;
                    ++version;
                    return true;
                }
            }
            return false;
        }

        public bool RemoveAll(T item)
        {
            bool isRemoved = false;
            T[] items = this.items;
            for (int i = 0, size = this.size; i < size; ++i)
            {
                if (EqualityComparer<T>.Default.Equals(item, items[i]))
                {
                    --size;
                    if (i < size)
                    {
                        Array.Copy(items, i + 1, items, i, size - i);
                    }
                    items[size] = default;
                    this.size = size;
                    isRemoved = true;
                }
            }
            if (isRemoved)
            {
                ++version;
                return true;
            }
            return false;
        }

        public ValueEnumerator GetEnumerator()
        {
            return new ValueEnumerator(this);
        }

        IFastEnumerator<T> IFastEnumerable<T>.GetEnumerator()
        {
            return new Enumerator(this);
        }

        private static void ThrowIndexOutOfRange()
        {
            throw new IndexOutOfRangeException("Access fault, Index is out of range.");
        }

        private static void ThrowExceedLimitedSize()
        {
            throw new OverflowException("Addition fault, Size is already reached the limited.");
        }

        private static void ThrowMismatchedVersion()
        {
            throw new InvalidOperationException("Iteration fault, Source element has been changed while iterating.");
        }

        IFastEnumerator IFastEnumerable.GetEnumerator()
        {
            return new Enumerator(this);
        }

        public ref struct ValueEnumerator
        {
            private OptimizedList<T> source;
            private int index, version;

            public ValueEnumerator(OptimizedList<T> source)
            {
                index = 0;
                this.source = source;
                version = source.version;
            }

            public void Dispose() { }

            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            public bool TryGetNext(out T next)
            {
                if (source.version != version)
                    ThrowMismatchedVersion();
                int i = index;
                if (i < source.size)
                {
                    index = i + 1;
                    next = source.items[i];
                    return true;
                }
                else
                {
                    next = default;
                    return false;
                }
            }
        }

        public sealed class Enumerator : IFastEnumerator<T>
        {
            private OptimizedList<T> source;
            private int index, version;

            public Enumerator(OptimizedList<T> source)
            {
                index = 0;
                this.source = source;
                version = source.version;
            }

            public void Dispose() { }

            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            public bool TryGetNext(out T next)
            {
                if (source.version != version)
                    ThrowMismatchedVersion();
                int i = index;
                if (i < source.size)
                {
                    index = i + 1;
                    next = source.items[i];
                    return true;
                }
                else
                {
                    next = default;
                    return false;
                }
            }

            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            public bool TryGetNext(out object next)
            {
                if (source.version != version)
                    ThrowMismatchedVersion();
                int i = index;
                if (i < source.size)
                {
                    index = i + 1;
                    next = source.items[i];
                    return true;
                }
                else
                {
                    next = default;
                    return false;
                }
            }
        }
    }
}
```

**Comment by YairHalberstadt** [comment]
Thanks @NutzNK that's much better.

Next precisely what language changes are you requesting?

Am I right they are:

1. Support for contravariant out parameters
2. Support for consuming IFastEnumerator in foreach

Do you think generator support is necessary as well?

I think my chief problem here is that the cost of enumerating is small enough that you need to make sure that IFastEnumerator is used everywhere to see much of a gain. Just adding it is not enough - you need to try to migrate existing code towards using it.

I tried to suggest such a path [here](https://github.com/dotnet/csharplang/issues/1931), but I don't think that would actually work.

In terms of improving Linq performance, I would suggest something closer to [this](https://github.com/dotnet/csharplang/issues/2482) would actually have a far bigger impact.

**Comment by NutzNK** [comment]
@YairHalberstadt 
1. Support for contravariant out parameters
    Yes, this supported function will make coding shorter like this part of above code
    public bool TryGetNext(out object next)
    will be able to coding like this
    public bool TryGetNext(out object next) => TryGetNext(out next)
    and IFastEnumerable<T> can change to IFastEnumerable<out T>

2. Support for consuming IFastEnumerator in foreach
    I cannot say "consuming IFastEnumerator" because foreach in c# in not binding to IEnumerable and IEnumerator but it has compilation rules and I want to modify foreach rule a little bit to full fill this request (it seem like Trait in rust language).
    Existing rules are
        1. Enumerable must contain method => TEnumerator GetEnumerator();
        2. TEnumerator from item 1. must contain Property => TValue Current{ get; } and Method => bool MoveNext();
    I want to change only 2nd rule to be
        2. TEnumerator from item 1. must contain Method => bool TryGetNext(out TValue value) or contain Property => TValue Current{ get; } and Method => bool MoveNext();

3. Both item 1st and 2nd need to modify in compiler.

4. In my opinion,  according to item 2 we don't want to implement IFastEnumerator ,
just Implement Method => bool TryGetNext(out TValue value) to IEnumerator interface which is already used for all enumerable is enough.
I declare IFastEnumerator for comparing with IEnumerator only.

5. Coding in your proposal cannot gain the best benefit yet,

Interface is
```C#
public interface IFastEnumerator<out T>
{
    T TryGetNext(out bool success);
}
```
then Iteration code behind will be like this
```C#
using (var enumerator = object.GetEnumerator())
{
    T value = enumerator.TryGetNext(out bool hasNext);
    while (hasNext)
    {
        //code in loop here
    }
}
```
but my code will iterate like this
```C#
using(var enumerator = object.GetEnumerator())
{
  while(enumerator.TryGetNext(out var value))
  {
    //code in loop here
  }
}
```
this code gain better performance on struct Enumerator

6. I understand it's not easy to implement this proposal because it will impact to several type that implement IEnumerable.

7. actually foreach in c# is free abstraction macro like C++ and Rust but it unable to coding, it's fixed in compiler. I understand this issue is not easy to change too because C# is base on runtime casting but C++ and Rust have compile time casting, I like some design from Rust and C++ too

**Comment by YairHalberstadt** [comment]
Note that out parameters don't actually exist in the runtime - they exist in C# only. From the runtime perspective they are just ordinary ref parameters. Thus this proposal would require runtime and language changes.

**Comment by NutzNK** [comment]
@YairHalberstadt 
yes, in and out is compile time checking, it compile to ref in runtime.
that why I think making out reference to support covariance type is not so much difficult,
because in instruction language
bool MoveNext() has similar instruction as void MoveNext(out bool result)
but these 2 method differences are inlining compilation and 1st method support covariance declaration but 2nd method doesn't.

**Comment by YairHalberstadt** [comment]
@NutzNK covariance is defined by the runtime, and ref in general is not covariant. So the runtime would need to recognize out, and make it covariant. I imagine that's not a small amount of work.

**Comment by NutzNK** [comment]
@YairHalberstadt ok, if it need to much work for work around this interface then I also have alternative way to do that but I'm not sure is it easier way to change or not.

use abstract instead of interface, it should be like this
```C#
public abstract class Enumerator
    {
        public abstract object Current { get; }

        public abstract bool MoveNext();
    }

    public abstract class Enumerator<T> : Enumerator, IDisposable
    {
        protected T current;
        public new T Current
        {
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            get => current;
        }

        public override object Enumerator.Current => current;

        public void Dispose() { }
    }
```
but problem is Abstract class still not be able to declare method like this "public override object Enumerator.Current => current;" this code will override only base class like interface.
this abstract class gain better performance than interface.

**Comment by CyrusNajmabadi** [comment]
i belive that the [shapes](https://github.com/dotnet/csharplang/issues/164) proposal would likely subsume this.

**Comment by NutzNK** [comment]
@CyrusNajmabadi  For me so many issues will be solved if C# can make same thing as "Trait" in Rust language.

**Comment by NutzNK** [comment]
@YairHalberstadt Now I found better way to do it.
I can improve more performance with current situation until new class can iterate faster than array but still has only 1 issue left, that is abstract class is not support co-variant type.

This is Benchmarking table

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18363
Intel Core i7-8750H CPU 2.20GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.1.101
[Host] : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT
DefaultJob : .NET Core 3.1.1 (CoreCLR 4.700.19.60701, CoreFX 4.700.19.60801), X64 RyuJIT

|                       Method |      Mean |    Error |   StdDev |       Min |       Max |    Median | Rank |
|----------------------------- |----------:|---------:|---------:|----------:|----------:|----------:|-----:|
|                Array_Foreach |  46.54 us | 0.203 us | 0.180 us |  46.26 us |  46.81 us |  46.56 us |    2 |
|    Array_IEnumerable_Foreach | 369.84 us | 1.087 us | 0.964 us | 369.01 us | 371.83 us | 369.27 us |    5 |
|                 List_Foreach | 197.94 us | 2.510 us | 2.225 us | 196.53 us | 203.59 us | 196.85 us |    3 |
|     List_IEnumerable_Foreach | 569.09 us | 2.468 us | 2.309 us | 566.31 us | 573.91 us | 568.65 us |    6 |
|          ExtremeList_Foreach |  40.75 us | 0.614 us | 0.574 us |  40.23 us |  42.14 us |  40.47 us |    1 |
| ExtremeList_IEnumerable_Foreach | 199.66 us | 0.475 us | 0.444 us | 199.14 us | 200.57 us | 199.46 us |    4 |

The key is these code
```C#
    public interface IEnumerator
    {
        object Current { get; }
        bool MoveNext();
    }

    public interface IEnumerator<out T> : IEnumerator
    {
        new T Current { get; }
    }

    public unsafe abstract class Enumerator : IEnumerator
    {
        protected Reference current;
        public object Current
        {
            [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
            get
            {
                return current.ToValue();
            }
        }

        public abstract bool MoveNext();
    }

    public unsafe abstract class Enumerator<T> : Enumerator, IEnumerator<T>
    {
        public new T Current
        {
            [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
            get
            {
                return current.ToValue<T>();
            }
        }
    }

    public interface IEnumerable
    {
        Enumerator GetEnumerator();
    }

    public interface IEnumerable<T>
    {
        Enumerator<T> GetEnumerator();
    }

    public static class EnumerableExtension
    {

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static IEnumerable<T> AsEnumerable<T>(this IEnumerable<T> source)
        {
            return source;
        }
    }
```
I create new struct `Reference` and `Reference<T>` that working at same way as `ByReference<T>` from microsoft.

And this is Enumerator inside

```C#
        public ref struct ValueEnumerator
        {
            private Reference<T> current, end;
            private Reference<int> source_version;
            private int version;

            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            public ValueEnumerator(ExtremeList<T> source)
            {
                current = new Reference<T>(ref source.items[0]);
                end = current + source.size;
                --current;
                source_version = new Reference<int>(ref source.version);
                version = source.version;
            }

            public T Current
            {
                [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
                get
                {
                    return current.Value;
                }
            }

            [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
            public bool MoveNext()
            {
                if (source_version.Value != version)
                    ThrowMismatchedVersion();
                ++current;
                if (current < end)
                {
                    return true;
                }
                --current;
                return false;
            }
        }

        public unsafe sealed class Enumerator : Enumerator<T>
        {
            private Reference end;
            private Reference<int> source_version;
            private int version;

            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            public Enumerator(ExtremeList<T> source)
            {
                var origin = Reference.New(ref source.items[0]);
                end = origin.Add<T>(source.size);
                current = origin.Previous<T>();
                source_version = new Reference<int>(ref source.version);
                version = source.version;
            }

            [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
            public override bool MoveNext()
            {
                if (source_version.Value != version)
                    ThrowMismatchedVersion();
                var next = current.Next<T>();
                if (next < end)
                {
                    current = next;
                    return true;
                }
                return false;
            }
        }
```

**Comment by ActivistInvestor** [comment]
Nice proposal, but I can't help but wonder about coupling MoveNext() and Current property into a single unit, because there are many cases in the framework, where MoveNext() is called repeatedly without any use of the Current item (e.g., Enumerable.Count(), Skip(), etc.). 

Since most of the LINQ iterators do a lot of work in MoveNext(), including getting the value of the Current property from the input/source, it may not matter much. However, I have diverged from that pattern, and use a lazy pattern for producing the value of the Current property, which has a noticeable effect on Count() and other cases where MoveNext() is repeatedly called without the use of the Current property.  The example above also uses this pattern to some extent, as it doesn't do any needless work in its MoveNext().

A distilled example of the pattern:

```
/// Overly-simplified pseudo-code

public class LazySelectIterator<TInput, TOutput> : IEnumerator<TOutput>
{
   bool _isvalid = false;
   Func<TInput, TOutput> _selector;
   IEnumerable<TInput> _source;
   IEnumerator<TInput> _enumerator;
   TOutput _current;

   public LazySelectIterator(IEnumerable<TInput> source, Func<TInput, TOutput> selector)
   {
      this._source = source;
      this._selector = selector;
   }

   public bool MoveNext()
   {
      if(_enumerator == null)
         _enumerator = _source.GetEnumerator();
      _isvalid = false;
      return _enumerator.MoveNext();
   }

   /// <summary>
   /// Defer any work required to produce the Current
   /// value until it's requested.
   /// </summary>

   public TOutput Current
   {
      get
      {
         if(!_isvalid)
         {
            _current = _selector(_enumerator.Current);
            _isvalid = true;
         }
         return _current;
      }
   }

   /// balance omitted for brevity
}

```

Have you tested the time of an equivalent of Count() on your prototype?
